packages/xod-client/src/user/actions.js

```diff
@@ -1,13 +1,65 @@
-import { getCompileLimitUrl } from '../utils/urls';
-import { UPDATE_COMPILE_LIMIT } from './actionTypes';
+import 'url-search-params-polyfill';
+
+import {
+  getCompileLimitUrl,
+  getWhoamiUrl,
+  getLoginUrl,
+  getLogoutUrl,
+} from '../utils/urls';
+import * as ActionTypes from './actionTypes';
 
 export const updateCompileLimit = (startup = false) => dispatch =>
   fetch(getCompileLimitUrl(), {
     headers: startup ? { 'x-launch': 'true' } : {},
   }).then(res => (res.ok ? res.json() : null))
     .catch(() => null)
     .then(limit => dispatch({
-      type: UPDATE_COMPILE_LIMIT,
+      type: ActionTypes.UPDATE_COMPILE_LIMIT,
       payload: limit,
     }));
+
+export const toggleAccountPane = () => ({
+  type: ActionTypes.TOGGLE_ACCOUNT_PANE,
+});
+
+const setGrant = grant => ({
+  type: ActionTypes.SET_AUTH_GRANT,
+  payload: grant,
+});
+
+export const getGrant = () => dispatch =>
+  fetch(getWhoamiUrl(), { credentials: 'include' })
+    .then(res => (res.ok ? res.json() : null))
+    .catch(() => null)
+    .then((grant) => {
+      dispatch(setGrant(grant));
+      dispatch(updateCompileLimit(false));
+    });
+
+export const login = (username, password) => (dispatch) => {
+  const form = new URLSearchParams();
+  form.set('username', username);
+  form.set('password', password);
+
+  dispatch({ type: ActionTypes.LOGIN_STARTED });
+
+  fetch(getLoginUrl(), { method: 'POST', credentials: 'include', body: form })
+    .then((res) => {
+      if (!res.ok) {
+        throw new Error(res);
+      }
+
+      return dispatch(getGrant());
+    })
+    .catch(() => dispatch({ type: ActionTypes.LOGIN_FAILED }));
```

Comment:
What if login service is unavailable, or network error occurred?
Maybe error should be passed into action payload?

P.S.
Also all these fetches, probably, should be wrapped with `retryOrFail` function from `xod-func-tools` to do retries on network/service unavailable errors...

Reply:
Oooh, very good point. Thought to test it in UI, but now it’s apparent network failures will not be handled gracefully.